torch/distributed/checkpoint/pg_aware_planner.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+import io
+from typing import Any, Dict, List
+
+import torch
+import torch.distributed.distributed_c10d as c10d
+from torch import distributed as dist
+from torch.distributed.checkpoint.default_planner import (
+    DefaultLoadPlanner,
+    DefaultSavePlanner,
+)
+from torch.distributed.checkpoint.metadata import (
+    STATE_DICT_TYPE,
+    Metadata,
+)
+from torch.distributed.checkpoint.planner import ReadItem
+from torch.distributed._shard.sharded_tensor.api import ShardedTensor
+
+
+def get_ranks(pg: c10d.ProcessGroup) -> List[int]:
+    """
+    Return an array of global ranks for a given process group.
+    """
+    return [c10d.get_global_rank(pg, i) for i in range(pg.size())]
+
+
+def create_state_dict_copy(state_dict: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Create the ``state_dict_copy`` used by ProcessGroupAwareSavePlanner and ProcessGroupAwareLoadPlanner.
+    It identifies ShardedTensor instances that are using the non-default
+    process groups and change their FQNs to be prefixed by a PG-specific string.
+    """
+    state_dict_copy = {}
+
+    for k, v in state_dict.items():
+        # Only rename the fqn if the current process group is a sub-process group.
+        if (
+            isinstance(v, ShardedTensor)
+            and v._process_group != dist.distributed_c10d._get_default_group()
+        ):
+            pg_global_ranks = get_ranks(v._process_group)
+            fqn = "_".join([str(rank) for rank in pg_global_ranks]) + "_" + k
+            state_dict_copy[fqn] = v
+        else:
+            state_dict_copy[k] = v
+
+    return state_dict_copy
+
+
+class ProcessGroupAwareSavePlanner(DefaultSavePlanner):
+    """
+    ProcessGroupAwareSavePlanner extends DefaultSavePlanner and re-write the state_dict.
```

Comment:
We should update the docstring to explain how this planner can be used.

Reply:
Sounds good. Will update in a separate PR.